File: dodgeGame/game.py
import pygame
import random
from .player import Player
from .enemy import Enemy
import math
from collections import Counter

# Set up the window
pygame.init()
pygame.display.set_caption("Dodger")

# Define colors
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
RED = (255, 0, 0)
ENEMY_NUM = 1

class Game:
    def __init__(self, window, width, height):
        self.window_width = width
        self.window_height = height
        self.window = window
        self.player = Player(self.window_width, self.window_height, RED)
        self.enemies = []
        self.score = [0]
        self.game_over = False
        self.game_over_wall = False
        self.clock = pygame.time.Clock()
        self.font = pygame.font.SysFont(None, 30)
        
        for i in range(ENEMY_NUM):
            self.enemies.append(Enemy(self.window_width, self.window_height))

    def handle_events(self):
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                # self.game_over = True
                pygame.quit()

        keys = pygame.key.get_pressed()
        self.player.move(keys)

    def move_enemies(self):
        for enemy in self.enemies:
            # enemy.move()
            enemy.move((self.player.x, self.player.y), self.score)

            # Check for collisions with the player
            if self.player.is_colliding_with(enemy):
                self.game_over = True

        # Increment the score
        # self.score = round(self.score + 0.001, 3)

    def draw(self):
        self.window.fill(BLACK)
        self.player.draw(self.window)
        for enemy in self.enemies:
            enemy.draw(self.window)
            pygame.draw.line(self.window, RED, (enemy.x, enemy.y), (self.player.x, self.player.y), 2)
        score_text = self.font.render(f"Score: {self.score[0]}", True, WHITE)
        self.window.blit(score_text, (10, 10))
        # print(self.player.last_locations, set(self.player.last_locations))
        # elemCounts = Counter(self.player.last_locations).values().sort(reverse=True)
        # if elemCounts[0] + elemCounts[1] > 15:
        #     self.game_over = True

    def run(self):
        while True:
            self.handle_events()
            self.move_enemies()
            self.draw()
            
            if self.player.touch_wall():
                self.game_over = True
            
            pygame.display.update()
            self.clock.tick(60)
            
            if self.game_over:
                self.reset()
                
            # print(self.get_state())
                
    def loop(self, move, numGames, numGamesDrawAfter=50000, scoreDrawAfter=10000):
        self.player.aiMove(move)
        self.move_enemies()
        
        if self.player.touch_wall():
            self.game_over_wall = True
        
        # if max(self.player.last_locations) - min(self.player.last_locations) < 20:
        #     self.game_over = True
        # if self.score > 0.01:
        #     print(self.player.last_locations, set(self.player.last_locations), len(set(self.player.last_locations)))
        # if self.score > 0.01 and len(set(self.player.last_locations)) < 4:
        #     self.game_over = True
        
        # if self.score[0] > 0.01:
        #     maxDist = 0
        #     for coord1 in self.player.last_locations:
        #         for coord2 in self.player.last_locations:
        #             if coord1 != coord2:
        #                 dist = math.sqrt((coord2[0] - coord1[0]) ** 2 + (coord2[1] - coord1[1]) ** 2)
        #                 if dist > maxDist:
        #                     maxDist = dist
        #     # print(self.player.last_locations, maxDist)
        #     if maxDist < 35:
        #         self.game_over = True
                
        blackScreen = True
        if numGames >= numGamesDrawAfter or self.score[0] >= scoreDrawAfter:
            self.draw()
            pygame.display.update()
            blackScreen = False
        if not blackScreen:
            self.window.fill(BLACK)
            blackScreen = True
        
        
    def reset(self):
        self.game_over = False
        self.game_over_wall = False
        self.player.reset()
        [enemy.reset() for enemy in self.enemies]
        self.score[0] = 0
        
    def get_state(self):
        state = self.player.getState()
        for enemy in self.enemies:
            state += ((enemy.x - self.player.x) / self.window_width, (enemy.y - self.player.y) / self.window_height)
            state += enemy.getState()
            # playerEnemyDir = math.atan2(self.player.y - enemy.y, self.player.x - enemy.x)
            # state += (((playerEnemyDir + 3.14) / 6.28),)
        return state

File: ui/game.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var player_1 = require("./player");
var enemy_1 = require("./enemy");
// import math
var DodgeGameEnv = /** @class */ (function () {
    function DodgeGameEnv(_a) {
        var _b = _a.render_mode, render_mode = _b === void 0 ? null : _b, _c = _a.window_size, window_size = _c === void 0 ? 64 : _c, _d = _a.model_window_size, model_window_size = _d === void 0 ? 64 : _d, _e = _a.enemy_movement, enemy_movement = _e === void 0 ? 'aimed' : _e, _f = _a.hp, hp = _f === void 0 ? 10 : _f, _g = _a.death_penalty, death_penalty = _g === void 0 ? 20 : _g, _h = _a.enemy_num, enemy_num = _h === void 0 ? 1 : _h, _j = _a.player_speed, player_speed = _j === void 0 ? 0.03 : _j, _k = _a.enemy_speed, enemy_speed = _k === void 0 ? 0.02 : _k, _l = _a.player_radius, player_radius = _l === void 0 ? 0.05 : _l, _m = _a.enemy_radius, enemy_radius = _m === void 0 ? 0.05 : _m, _o = _a.action_space, action_space = _o === void 0 ? 4 : _o, _p = _a.normalize, normalize = _p === void 0 ? true : _p, _q = _a.randomize_player_speed, randomize_player_speed = _q === void 0 ? false : _q, _r = _a.randomize_enemy_speed, randomize_enemy_speed = _r === void 0 ? false : _r, _s = _a.randomize_player_radius, randomize_player_radius = _s === void 0 ? false : _s, _t = _a.randomize_enemy_radius, randomize_enemy_radius = _t === void 0 ? false : _t, _u = _a.randomize_enemy_num, randomize_enemy_num = _u === void 0 ? false : _u;
        this.metadata = {
            render_modes: ['human', 'rgb_array'],
            render_fps: 30,
        };
        this.enemies = [];
        this.score = 0;
        this.game_over = false;
        this.clock = null;
        this.window = null;
        this.canvas = null;
        this.font = "30px sans-serif";
        // Define colors
        this.BLACK = '#000000';
        this.WHITE = '#FFFFFF';
        this.RED = '#FF0000';
        this.model_window_size = model_window_size;
        this.window_size = window_size;
        this.player = new player_1.default(this.window_size, this.window_size, this.WHITE, player_radius, player_speed, action_space, randomize_player_radius, randomize_player_speed, normalize);
        this.score = 0;
        this.game_over = false;
        this.enemy_movement = enemy_movement;
        this.max_enemy_num = enemy_num;
        this.randomize_enemy_num = randomize_enemy_num;
        this.enemy_num = this.randomize_enemy_num
            ? Math.floor(Math.random() * (this.max_enemy_num - 1) + 1)
            : this.max_enemy_num;
        this.enemy_speed = enemy_speed;
        this.enemy_radius = enemy_radius;
        this.randomize_enemy_speed = randomize_enemy_speed;
        this.randomize_enemy_radius = randomize_enemy_radius;
        this.max_hp = hp;
        this.hp = hp;
        this.death_penalty = death_penalty;
        this.normalize = normalize;
        this.canvas = document.getElementById('canvas');
        this.canvas.width = this.window_size;
        this.canvas.height = this.window_size;
        this.window = this.canvas.getContext("2d");
        for (var i = 0; i < this.enemy_num; i++) {
            this.enemies.push(new enemy_1.default(this.window_size, this.window_size, this.RED, enemy_speed, enemy_radius, enemy_movement, randomize_enemy_radius, randomize_enemy_speed, normalize));
        }
    }
    DodgeGameEnv.prototype.getObservation = function () {
        return this.window.getImageData(0, 0, this.canvas.width, this.canvas.height).data;
    };
    DodgeGameEnv.prototype.reset = function () {
        this.game_over = false;
        this.player.reset();
        this.enemies.forEach(function (enemy) { return enemy.reset(); });
        this.score = 0;
        this.hp = this.max_hp;
        // this.window_size = randint(64, 128);
        if (this.randomize_enemy_num) {
            this.enemy_num = Math.random() * (this.max_enemy_num - 1) + 1;
            this.enemies = [];
            // console.log(this.enemy_num);
            for (var i = 0; i < this.enemy_num; i++) {
                this.enemies.push(new enemy_1.default(this.window_size, this.window_size, this.RED, this.enemy_speed, this.enemy_radius, this.enemy_movement, this.randomize_enemy_radius, this.randomize_enemy_speed, this.normalize));
            }
        }
        var observation = this.getObservation();
        return observation;
    };
    DodgeGameEnv.prototype.isGameOver = function () {
        var _this = this;
        return this.enemies.some(function (enemy) {
            var distance = Math.sqrt(Math.pow((_this.player.pos.x - enemy.pos.x), 2) + Math.pow((_this.player.pos.y - enemy.pos.y), 2));
            return distance < _this.player.radius + enemy.radius;
        }) || this.player.pos.x < 0 || this.player.pos.x > 1 || this.player.pos.y < 0 || this.player.pos.y > 1;
    };
    DodgeGameEnv.prototype.step = function (action) {
        if (this.game_over) {
            this.game_over = false;
        }
        this.player.aiMove(action);
        this.moveEnemies();
        if (this.player.touchWall()) {
            this.game_over = true;
        }
        var gameOver = this.isGameOver();
        this.hp -= gameOver ? 1 : 0;
        var reward = 1 + (this.game_over ? -1 * this.death_penalty : 0);
        this.score += reward;
        var observation = this.getObservation();
        var done = this.hp <= 0;
        var info = {};
        return [observation, reward, done, info];
    };
    DodgeGameEnv.prototype.render = function () {
        var _this = this;
        this.window.clearRect(0, 0, this.window_size, this.window_size);
        this.player.draw(this.window);
        this.enemies.forEach(function (enemy) { return enemy.draw(_this.window); });
        this.window.font = this.font;
        this.window.fillStyle = this.BLACK;
        this.window.fillText("Score: ".concat(this.score), 10, 50);
        this.window.fillText("HP: ".concat(this.hp), 10, 80);
    };
    DodgeGameEnv.prototype.moveEnemies = function () {
        for (var _i = 0, _a = this.enemies; _i < _a.length; _i++) {
            var enemy = _a[_i];
            enemy.move(this.player.pos);
            // Check for collisions with the player
            if (this.player.isCollidingWith(enemy)) {
                this.game_over = true;
            }
        }
    };
    DodgeGameEnv.prototype.run = function () {
        this.reset();
        this.render();
        this.game_over = false;
        while (true) {
            var action = Math.floor(Math.random() * 4);
            this.step(action);
            this.render();
        }
    };
    return DodgeGameEnv;
}());
exports.default = DodgeGameEnv;
